src/Orchard.Environment.Extensions/ExtensionManager.cs

```diff
@@ -307,5 +321,22 @@ private static string GetSourceFeatureNameForType(Type type, string extensionId)
             }
             return extensionId;
         }
+
+        private class LazyConcurrentDictionary<TKey, TValue>
+        {
+            private readonly ConcurrentDictionary<TKey, Lazy<TValue>> concurrentDictionary;
+
+            public LazyConcurrentDictionary()
+            {
+                this.concurrentDictionary = new ConcurrentDictionary<TKey, Lazy<TValue>>();
+            }
+
+            public TValue GetOrAdd(TKey key, Func<TKey, TValue> valueFactory)
+            {
+                var lazyResult = this.concurrentDictionary.GetOrAdd(key, k => new Lazy<TValue>(() => valueFactory(k), LazyThreadSafetyMode.ExecutionAndPublication));
```

Comment:
Why the thread safety option? Either you comment the explicit reason because it is understood and it solves an issue, or you use the default.

Also, can we write it `new Lazy<TValue>(valueFactory)` ?